### ISSUE TITLE
HttpClient extension methods causes ambiguous invocations in .NET 7

### ISSUE THREAD

**Issue by siewers** [WireMock-Net/WireMock.Net#974]
### Describe the bug
When adding WireMock to a .NET 7 project, it breaks the usage of certain HttpClient extension methods by causing ambiguous invocations.

### Expected behavior:
There should be no ambiguous invocations/conflicts when using WireMock.

### Test to reproduce
1. Create a new .NET 7 class library.
2. Add the following to the Class1.cs file:
```cs
using System.Net.Http.Json;

namespace ClassLibrary1;

public class Class1
{
    public void Foo()
    {
        var httpClient = new HttpClient();
        httpClient.PostAsJsonAsync("test", new { Foo = "bar" }, CancellationToken.None);
    }
}
```
3. Build the project and confirm there are no compilation errors.
4. Add a package reference to WireMock (at this time the latest version is 1.5.32).
```xml
<ItemGroup>
    <PackageReference Include="WireMock.Net" Version="1.5.32" />
</ItemGroup>
```
5. Build the project and observe the compilation error about the ambiguous invocation on `PostAsJsonAsync`.

### Other related info
There might be other extension methods that cause the same issue.
Removing the cancellation token from the `PostAsJsonAsync` call will resolve the extension method located in the seemingly deprecated [`System.Net.Http.Formatting`](https://learn.microsoft.com/en-us/previous-versions/aspnet/hh944845(v=vs.118)) assembly, which seems to originate from the `Microsoft.AspNet.WebApi.Client` package according to Visual Studio.
![image](https://github.com/WireMock-Net/WireMock.Net/assets/226147/f9634214-356f-484f-b154-a637898a30ea)

This resolves to `System.Net.Http.HttpClientExtensions` (System.Net.Http.Formatting.dll):
```cs
httpClient.PostAsJsonAsync("test", new { Foo = "bar" });
```
This resolves to `System.Net.Http.Json.HttpClientJsonExtensions` (System.Net.Http.Json.dll):
```cs
httpClient.PostAsJsonAsync("test", new { Foo = "bar" }, JsonSerializerOptions.Default, CancellationToken.None);
```

### Potential workaround
To resolve the ambiguity, it is possible to invoke the extension method as a static method instead, but this is not ideal and will require you to remember this issue in order to not end up in this situation again.
```cs
HttpClientJsonExtensions.PostAsJsonAsync(httpClient, "test", new { Foo = "bar" }, CancellationToken.None);
```
Another solution is to give the `System.Net.Http.Formatting` assembly a different alias (see this [StackOverflow answer](https://stackoverflow.com/a/52252705/76048))
```xml
<Target Name="ChangeAliasOfSystemNetHttpFormatting" BeforeTargets="FindReferenceAssembliesForReferences;ResolveReferences">
  <ItemGroup>
    <ReferencePath Condition="'%(FileName)' == 'System.Net.Http.Formatting'">
      <Aliases>nonmerged</Aliases>
    </ReferencePath>
  </ItemGroup>
</Target>
```
This is probably the nicest way to solve it, but it's still something that's less than ideal.

**Comment by StefH** [comment]
@siewers 
I applied your last suggestion.

Can you review PR https://github.com/WireMock-Net/WireMock.Net/pull/977 ?

**Comment by StefH** [comment]
Can you try preview version 1.5.32-ci-17648 ?

https://github.com/WireMock-Net/WireMock.Net/wiki/MyGet-preview-versions

**Comment by siewers** [comment]
Unfortunately, that won't work.
The reason is that the package is transitively available to all consumers of WireMock.Net. What this change does is only aliasing it within the WireMock.Net project, but not for consumers.
Since you aren't referencing the package directly in WireMock, it's not possible to hide that one directly from consumers, but perhaps there is a trick to mark it with `PrivateAssets="compile;build;analyzers"` somehow, so it won't be accessible to consumers.
It might even be possible to hide the `Microsoft.AspNet.WebApi.Client` package reference like so:
```xml
<PackageReference Include="Microsoft.AspNet.WebApi.Client" Version="5.2.8" PrivateAssets="compile;build;analyzers" />
```
I believe (but I'm not sure) that hiding a package reference like this also hides all its transitive dependencies, but I would need to test it.

In general, I think WireMock exposes too many of its internal dependencies, resulting in IntelliSense becoming extremely bloated with all kinds of references not included in the consuming project itself.
I believe it's good practice to hide the library dependencies to consumers to not unnecessarily bloat the IntelliSense and accidentally start relying on a dependency not directly referenced (e.g. the GraphQL package that exposes a lot of extension methods as well).
If you ever decide to remove or upgrade such a dependency, you might inadvertently break consumers of WireMock, because they in turn relied on some random extension method from a package you included, which is now removed or has changed.

I know there are probably a lot of opinions in this area, but it's risky exposing dependencies not intended to be used by consumers.

On the other hand, there's also the risk of package version mismatch, if the consumer directly references a package that WireMock.Net depends on, but hides. In this case, one way to solve it could be to add a [version range](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges) in the package reference, allowing consumers to depend on e.g. `Newtonsoft.Json` version 12 through 18 or something like that. If the consumer upgrades the package to version 19, the project won't build because WireMock.Net isn't tested or guaranteed compatible with that version.

It's a lot of complexity for such a seemingly tiny problem, but it's definitely worth considering for future maintainability.

Oh, one more thing, _if_ you decide to hide all internal dependencies, it's most likely a major breaking change, as nobody knows what packages are being used transitively by consumers.

**Comment by StefH** [comment]
Thank you for your detailed explanation and suggestions.
I think that marking the dependencies as PrivateAssets is the best solution in the long term which indeed could mean a breaking change.

(https://learn.microsoft.com/en-us/answers/questions/702182/how-to-hide-dependent-nuget-dlls-from-consuming-pa)

For now I've deleted the PR.

Another approach which I was trying to build for this specific issue is removing the dependency on Microsoft.AspNet.WebApi.Client (which I use for creating a httpclient by using a httpclientfactory) for .NET 6 and higher. I did build a work-around using code like this:
``` c#
public class X
{
	public static X Instance = new X();
	
	public readonly IHttpClientFactory Factory;
	
	private X()
	{
		var services = new ServiceCollection();
		services.AddHttpClient("WireMock.Net.HttpClient");
		
		Factory = services.BuildServiceProvider().GetRequiredService<IHttpClientFactory>();
	}
}
```

This worked, however the current code uses:
``` c#
public HttpClient CreateClient(params DelegatingHandler[] handlers)
```

And the handlers are not supported in that same way....

I need to think on that some more...

**Comment by siewers** [comment]
You may want to have a look at the documentation here https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory#configure-the-httpmessagehandler

The pattern is to inject a typed, configured, client into `X` and then use that one instead.

I haven’t checked for .NET 6 and above, but at some point there was a static factory method on the ‘HttpClientFactory` class, I’m just not sure if that exists anymore.

**Comment by StefH** [comment]
I was also checking this 
https://github.com/myarichuk/Simple.HttpClientFactory

**Comment by siewers** [comment]
Well, it's an option, although I'm having a hard time finding it justifiable today.
Sure, you save a few lines of internal implementation code, but the project seems stale and haven't had an update for three years and only 8 stars. The rationale behind the project seems to be a bit of a stretch IMHO :)
I'd suggest sticking with the framework and not pull in other third-party packages to solve this problem :)

**Comment by siewers** [comment]
I had a look at the implementation in the [`HttpClientFactory`](https://github.com/aspnet/AspNetWebStack/blob/main/src/System.Net.Http.Formatting/HttpClientFactory.cs) and I believe this should be able to do the same without the need for the package dependency:
```cs
using System.Linq;
using System.Net.Http;

namespace WireMock.Http;

internal static class HttpClientFactory2
{
    public static HttpClient Create(params DelegatingHandler[] handlers)
    {
        var handler = CreateHandlerPipeline(new HttpClientHandler(), handlers);
        return new HttpClient(handler);
    }

    public static HttpClient Create(HttpMessageHandler innerHandler, params DelegatingHandler[] handlers)
    {
        var handler = CreateHandlerPipeline(innerHandler, handlers);
        return new HttpClient(handler);
    }

    private static HttpMessageHandler CreateHandlerPipeline(HttpMessageHandler handler, params DelegatingHandler[] delegatingHandlers)
    {
        if (delegatingHandlers.Length == 0)
        {
            return handler;
        }

        var next = handler;
        
        foreach (var delegatingHandler in delegatingHandlers.Reverse())
        {
            delegatingHandler.InnerHandler = next;
            next = delegatingHandler;
        }

        return next;
    }
}
```

**Comment by StefH** [comment]
https://github.com/WireMock-Net/WireMock.Net/pull/996

**Comment by StefH** [comment]
Can you try preview: 1.5.32-ci-17765

https://github.com/WireMock-Net/WireMock.Net/wiki/MyGet-preview-versions

**Comment by siewers** [comment]
I can't find the build on MyGet. The only prerelease versions available are 1.5.34 and 1.5.35.

**Comment by StefH** [comment]
Did you select the preview select box in visual studio?
